disperser/apiserver/server.go

```diff
@@ -303,6 +304,107 @@ func (s *DispersalServer) disperseBlob(ctx context.Context, blob *core.Blob, aut
 	}, nil
 }
 
+func (s *DispersalServer) DispersePaidBlob(ctx context.Context, req *pb.DispersePaidBlobRequest) (*pb.DisperseBlobReply, error) {
+	blob, err := s.validatePaidRequestAndGetBlob(ctx, req)
+	quorumNumbers := req.CustomQuorumNumbers
+	binIndex := req.PaymentHeader.BinIndex
+	cumulativePayment := new(big.Int).SetBytes(req.PaymentHeader.CumulativePayment)
+	//todo: before disperse blob, validate the signature
+	signature := req.PaymentSignature
+
+	if err != nil {
+		for _, quorumID := range req.CustomQuorumNumbers {
+			s.metrics.HandleFailedRequest(codes.InvalidArgument.String(), fmt.Sprint(quorumID), len(req.GetData()), "DispersePaidBlob")
+		}
+		s.metrics.HandleInvalidArgRpcRequest("DispersePaidBlob")
+		return nil, api.NewInvalidArgError(err.Error())
+	}
+
+	reply, err := s.dispersePaidBlob(ctx, blob, quorumNumbers, binIndex, cumulativePayment, signature, "", "DispersePaidBlob")
+	if err != nil {
+		// Note the DispersePaidBlob already updated metrics for this error.
+		s.logger.Info("failed to disperse blob", "err", err)
+	} else {
+		s.metrics.HandleSuccessfulRpcRequest("DispersePaidBlob")
+	}
+	return reply, err
+}
+
+// dispersePaidBlob checks for payment metering, otherwise does the same thing as disperseBlob
+func (s *DispersalServer) dispersePaidBlob(ctx context.Context, blob *core.Blob, quorumNumbers []uint32, binIndex uint32, cumulativePayment *big.Int, signature []byte, authenticatedAddress string, apiMethodName string) (*pb.DisperseBlobReply, error) {
```

Comment:
looks like its mostly copied from `disperseBlob` except for the `s.meterer != nil` block. Why don't we integrate it into the existing method?

Reply:
I didn't dare mess around `disperseBlob` 😬 and wanted to keep the two flows cleanly separated from each other. Additionally, `dispersePaidBlob` has these extra params ` quorumNumbers []uint32, binIndex uint32, cumulativePayment *big.Int, signature []byte`.

To make the code cleaner, I could add a `PaymentMetadata` param to `disperseBlob` and pass in an empty one when calling from `DisperseBlob` and construct a real one in `DispersePaidBlob`. What do you think?

Reply:
> To make the code cleaner, I could add a PaymentMetadata param to disperseBlob and pass in an empty one when calling from DisperseBlob and construct a real one in DispersePaidBlob. What do you think?

I'd be in favor of something like this to reduce code duplication

Reply:
updated! removed `dispersePaidBlob` and added `PaymentMetadata` to `disperseBlob`